src/main/java/com/mongodb/hibernate/jdbc/MongoPreparedStatement.java

```diff
@@ -59,15 +71,35 @@
  */
 final class MongoPreparedStatement extends MongoStatement implements PreparedStatementAdapter {
 
-    private final BsonDocument command;
+    private final List<BsonDocument> commandBatch;
+    private final @Nullable BsonDocument commandPrototype;
+
+    private BsonDocument command;
 
     private final List<Consumer<BsonValue>> parameterValueSetters;
 
-    public MongoPreparedStatement(
+    MongoPreparedStatement(
             MongoClient mongoClient, ClientSession clientSession, MongoConnection mongoConnection, String mql) {
+        this(mongoClient, clientSession, mongoConnection, mql, false);
+    }
+
+    MongoPreparedStatement(
+            MongoClient mongoClient,
+            ClientSession clientSession,
+            MongoConnection mongoConnection,
+            String mql,
+            boolean batchable) {
         super(mongoClient, clientSession, mongoConnection);
-        this.command = BsonDocument.parse(mql);
-        this.parameterValueSetters = new ArrayList<>();
+        if (batchable) {
```

Comment:
The `batchable` parameter and this conditional logic should be removed.  Just because batching is global enabled doesn't mean it's actually going to be used in this particular statement, which I think you can only determine once `#addBatch` is called.  So whatever initialization is done in the constructor has to work for both cases.

If this is being done as a performance optimization, it is a premature one.

Reply:
This might be one of the best code review comments I ever received. I removed its usage and found this PR was simplified so much!
Yeah, it is premature perf opt (to save some command prototype creation, etc.).